src/Features/CSharp/Portable/CodeRefactorings/InlineMethod/CSharpInlineMethodRefactoringProvider.cs

```diff
@@ -0,0 +1,157 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.CodeGeneration;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.LanguageServices;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.InlineMethod;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.InlineMethod
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(PredefinedCodeRefactoringProviderNames.InlineMethod)), Shared]
+    [Export(typeof(CSharpInlineMethodRefactoringProvider))]
+    internal sealed class CSharpInlineMethodRefactoringProvider : AbstractInlineMethodRefactoringProvider
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CSharpInlineMethodRefactoringProvider() : base(CSharpSyntaxFacts.Instance)
+        {
+        }
+
+        protected override async Task<SyntaxNode?> GetInvocationExpressionSyntaxNodeAsync(CodeRefactoringContext context)
+        {
+            var syntaxNode = await context.TryGetRelevantNodeAsync<InvocationExpressionSyntax>().ConfigureAwait(false);
+            return syntaxNode;
+        }
+
+        private static bool ShouldStatementBeInlined(StatementSyntax statementSyntax)
+            => statementSyntax is ReturnStatementSyntax || statementSyntax is ExpressionStatementSyntax;
+
+        protected override bool IsMethodContainsOneStatement(SyntaxNode calleeMethodDeclarationSyntaxNode)
+        {
+            if (calleeMethodDeclarationSyntaxNode is MethodDeclarationSyntax declarationSyntax)
```

Comment:
we likely want to support anything that has a body.  so accessors, operators, cosnturctors, etc.

Reply:
What I saw @genlu mentioned in #22052 is 
"only allowed on private method with simple body (single statement body and expression body)"
I didn't attend the original design meeting so I am not sure if operators/accessors should be included.

The current implementation only contains 'method' inclining. 
I will leave this comment open and wait for Gen's suggestions.

Reply:
we shoudl only inline methods.  However, taht method might be *called* from otehr contxts (like a property accessor).  We should allow inlining from any calling context.

Reply:
I think `calleeMethodDeclarationSyntaxNode` is only accept the callee's method declaration. 
Why will this affect the calling context?

Reply:
ah, this is the callee.  sorry, i thought this was the caller.  

do yo have tests that you can inline a method call from something like a constructor just to verify this works?

Reply:
I have some test in the property. I will modify some them to be called from ctor & other places